lld/ELF/ICF.cpp

```diff
@@ -286,9 +313,14 @@ bool ICF<ELFT>::constantEq(const InputSection *secA, Relocs<RelTy> ra,
     // Relocations referring to InputSections are constant-equal if their
     // section offsets are equal.
     if (isa<InputSection>(da->section)) {
-      if (da->value + addA == db->value + addB)
+      if (da->value + addA == db->value + addB) {
```

Comment:
This change was not in #134342 . I have tried to understand why the logic is so complex.
Looks like https://reviews.llvm.org/D34094 , while supporting MergeInputSection, also introduced the untested condition for InputSection
```
    if (isa<InputSection>(DA->Section))
      return DA->Value + AddA == DB->Value + AddB;
```

I feel that we should switch to the safer `da->value == db->value && addA == addB`, which likely doesn't lose anything (I believe it's difficult for clang emitted code to utilize the relaxed condition.)

Reply:
I think `da->value + addA == db->value + addB` is more readable. I can read it as if the address resolves to the same place, sections are considered same. It also takes care of section folding for more cases. Changing it to `da->value == db->value && addA == addB` would exclude some cases.

Reply:
> Changing it to `da->value == db->value && addA == addB` would exclude some cases.

The excluded cases are likely negligible. This condition is for InputSection (non merge string), where symbols defined not at offset 0 in -ffunction-sections -fdata-sections code are extremely rare. I don't know a case LLVM codegen would do this.

Reply:
The chromium build that made us revert the original change did have many sections that were being merged due to this property and @zmodem tried my second commit in this PR (https://github.com/llvm/llvm-project/pull/139493/commits/e4041b71b7c2f0abf2bc0baa926cccd0ab74bf55) separately and noticed a binary size regression that was unacceptable to them. This comment on that Chromium bug is relevant: https://b.corp.google.com/issues/415810137#comment44

Reply:
If you insist, I propose to do that in a separate PR so that these changes are not affected by any potential revert.

Reply:
If it helps break disagreement, I lean towards @MaskRay 's suggestion to use `da->value == db->value && addA == addB` as the condition (same value, same addend). Maybe it misses something, but it's conceptually simpler even if it misses some case.

Reply:
I corrected the link in my previous comment to a public one. Apologies I didn't notice that I pasted an internal link to the Chromium bug. 

Using `da->value == db->value && addA == addB` may seem trivial but it misses cases that increases the size of chromium builds to unacceptable levels. As I mentioned in my previous comment, @zmodem had tried it already and noticed a large size regression.

Reply:
Hmm. Thanks for clarification. I was under the impression that after my comment here (https://issues.chromium.org/issues/415810137#comment50), you tried it and confirmed it works but the size regression was not acceptable to you. 

I am okay simplifying the condition here after this clarification.